File: benchmarks/62-arbitrary-precision/arbitrary_precision.py
#!/usr/bin/env python3
"""Experiment 62 -- Arbitrary-precision numerics (mpmath column).

Same kernels as ``arbitrary_precision.m``, same order and labels.

Baseline is mpmath -- the standard Python arbitrary-precision library (scipy is
double-only, so it cannot appear here).  mpmath is a weaker baseline than scipy
in one sense: pure Python unless gmpy2 is installed, in which case its core is
GMP/MPFR -- the same library Mathilda's precision path uses.  The .py docstring
of a case says which.

Single-run (bench_once): these computations are expensive on their own.  Checks
use a deep 6-digit window win(v, s) = digits s-5..s of v, matching the .m side.
"""

import sys, os; sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

import mpmath as mp

from harness import bench_once, check, require

require(["mpmath:pi", "mpmath:nsum", "mpmath:quad", "mpmath:findroot"])


def win(v, start):
    return int(mp.floor(v * mp.mpf(10) ** start)) % 10 ** 6


def n_pi():
    mp.mp.dps = 10020
    return +mp.pi

bench_once("N[Pi, 10000]", n_pi)
check("N[Pi, 10000]", win(n_pi(), 5000))


def n_gamma():
    mp.mp.dps = 1020
    return mp.gamma(mp.mpf(1) / 3)

bench_once("N[Gamma[1/3], 1000]", n_gamma)
check("N[Gamma[1/3], 1000]", win(n_gamma(), 500))


def n_log2():
    mp.mp.dps = 5020
    return mp.log(2)

bench_once("N[Log[2], 5000]", n_log2)
check("N[Log[2], 5000]", win(n_log2(), 2000))


def s_zeta2():
    mp.mp.dps = 120
    return mp.nsum(lambda k: 1 / k ** 2, [1, mp.inf])

bench_once("NSum 1/k^2 WorkingPrecision 100", s_zeta2)
check("NSum 1/k^2 WorkingPrecision 100", win(s_zeta2(), 60))


def i_gauss():
    mp.mp.dps = 220
    return mp.quad(lambda x: mp.e ** (-x * x), [0, 1])

bench_once("NIntegrate Exp[-x^2] WorkingPrecision 200", i_gauss)
check("NIntegrate Exp[-x^2] WorkingPrecision 200", win(i_gauss(), 50))


def r_dottie():
    mp.mp.dps = 220
    return mp.findroot(lambda x: mp.cos(x) - x, 1)

bench_once("FindRoot Cos[x]=x WorkingPrecision 200", r_dottie)
check("FindRoot Cos[x]=x WorkingPrecision 200", win(r_dottie(), 50))

File: benchmarks/62-arbitrary-precision/arbitrary_precision.m
(* ==========================================================================
   Experiment 62 -- Arbitrary-precision numerics (group E)
   ==========================================================================
   WHAT IT MEASURES.  The MPFR-backed high-precision path: N[expr, p] and the
   WorkingPrecision option honoured by NSum, NIntegrate and FindRoot
   (src/precision.c and the *_mpfr.c modules).  This is the one group-E
   experiment whose baseline is NOT scipy -- scipy is double-only -- but mpmath,
   the standard Python arbitrary-precision library.  A README backlog item.

   Single-run (benchOnce): N[Pi, 10000] and friends are expensive on their own,
   and a warm-up would let Mathilda's N[] result cache serve every later rep for
   free -- measuring the cache, not the arithmetic.

   Checks use a 6-digit window pulled from DEEP in the expansion
   (win[v, s] = digits s-5..s of v), so the check verifies real precision, not
   just the leading figures.  FindRoot at WorkingPrecision -> 200 now delivers
   ~wp/2 = 100 correct digits (win50); it previously stopped at ~19 digits while
   labelling the result Precision 100, a correctness bug fixed by scaling
   AccuracyGoal with WorkingPrecision (findroot.c).  NSum at WorkingPrecision ->
   100 now reaches ~90+ digits (win60); it previously floored at ~44 because its
   Euler–Maclaurin base did not scale with precision (nsum.c).
   ========================================================================== *)

Get["../harness.m"];
Get["../data.m"];

require[{"N", "NSum", "NIntegrate", "FindRoot", "WorkingPrecision"}];

win[v_, start_] := Mod[IntegerPart[v 10^start], 10^6];

benchOnce["N[Pi, 10000]", N[Pi, 10000];];
check["N[Pi, 10000]", win[N[Pi, 10000], 5000]];

benchOnce["N[Gamma[1/3], 1000]", N[Gamma[1/3], 1000];];
check["N[Gamma[1/3], 1000]", win[N[Gamma[1/3], 1000], 500]];

benchOnce["N[Log[2], 5000]", N[Log[2], 5000];];
check["N[Log[2], 5000]", win[N[Log[2], 5000], 2000]];

benchOnce["NSum 1/k^2 WorkingPrecision 100",
          NSum[1/k^2, {k, 1, Infinity}, WorkingPrecision -> 100];];
check["NSum 1/k^2 WorkingPrecision 100",
      win[NSum[1/k^2, {k, 1, Infinity}, WorkingPrecision -> 100], 60]];

benchOnce["NIntegrate Exp[-x^2] WorkingPrecision 200",
          NIntegrate[Exp[-x^2], {x, 0, 1}, WorkingPrecision -> 200];];
check["NIntegrate Exp[-x^2] WorkingPrecision 200",
      win[NIntegrate[Exp[-x^2], {x, 0, 1}, WorkingPrecision -> 200], 50]];

benchOnce["FindRoot Cos[x]=x WorkingPrecision 200",
          FindRoot[Cos[x] == x, {x, 1}, WorkingPrecision -> 200];];
check["FindRoot Cos[x]=x WorkingPrecision 200",
      win[x /. FindRoot[Cos[x] == x, {x, 1}, WorkingPrecision -> 200], 50]];
